### PR TITLE
Add unions in core types

### PR DESCRIPTION
**Pull request by apelisse**
**What type of PR is this?**
/kind api-change
/kind feature

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes #35345, and related to #35343, #24238, #34292, #6979, #33766, #24198, #60340 and others.

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
Core API types now support, understand union semantics, and try to determine how to handle changes to union fields.
```

### CODE REVIEW

**Comment by cblecker** [comment]
This bumps a whole bunch of dependencies, and creates deviation between the pinned and preferred versions for a number of deps. We need to look at this one closely.

cc: @liggitt

**Comment by apelisse** [comment]
> This bumps a whole bunch of dependencies, and creates deviation between the pinned and preferred versions for a number of deps. We need to look at this one closely.

Yeah, not sure why it did that, guidance is welcome :-)

**Comment by liggitt** [comment]
> Yeah, not sure why it did that, guidance is welcome :-)

what versions of kube-openapi and structured-merge-diff are you wanting to update to? what commands did you run to build the `Update go-openapi and structured-merge-diff` commit?

I think newer versions of go-openapi pull in a lot of changes :-/

**Comment by liggitt** [comment]
> I think newer versions of go-openapi pull in a lot of changes :-/

I take it back, looks like the module you're bumping has identical dependencies at 0.17.2 and 0.19.0, which is good

**Comment by apelisse** [comment]
I think the command I used were:
```
go get -u k8s.io/kube-openapi
go get -u sigs.k8s.io/structured-merge-diff
# Manually change the replace lines in go.mod
```

**Comment by liggitt** [comment]
ah, see https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/vendor.md#adding-or-updating-a-dependency for the steps to update. 

doing this gave a smaller diff:
```
hack/pin-dependency.sh github.com/go-openapi/spec v0.19.0
hack/pin-dependency.sh k8s.io/gengo v0.0.0-20190128074634-0689ccc1d7d6
hack/pin-dependency.sh k8s.io/kube-openapi v0.0.0-20190502190224-411b2483e503
```

it still left new `replace` statements in some of the staging components' go.mod files, which we don't want. it looks like this bumps from a level of kube-openapi without a go.mod file (https://github.com/kubernetes/kube-openapi/blob/b3a7cee44a30/go.mod) to one with one (https://github.com/kubernetes/kube-openapi/blob/411b2483e503/go.mod), which is confusing the [pruning bits](https://github.com/kubernetes/kubernetes/blob/master/hack/update-vendor.sh#L251-L264) of hack/update-vendor.sh. I'll see if I can get that bit cleaned up.

**Comment by apelisse** [comment]
Yep, I've clearly added `go.mod` in kube-openapi

File: go.mod [hunk 1]
```diff
@@ -468,7 +468,7 @@ replace (
 	modernc.org/strutil => modernc.org/strutil v1.0.0
 	modernc.org/xc => modernc.org/xc v1.0.0
 	sigs.k8s.io/kustomize => sigs.k8s.io/kustomize v2.0.3+incompatible
-	sigs.k8s.io/structured-merge-diff => sigs.k8s.io/structured-merge-diff v0.0.0-20190302045857-e85c7b244fd2
+	sigs.k8s.io/structured-merge-diff => sigs.k8s.io/structured-merge-diff v0.0.0-20190302045857-ea680f03cc65b
```
Comment (liggitt) [inline go.mod]: this isn't quite new enough. looks like `k8s.io/kube-openapi@v0.0.0-20190502190224-411b2483e503` requires `sigs.k8s.io/structured-merge-diff@v0.0.0-20190426204423-ea680f03cc65`

File: go.mod [hunk 2]
```diff
@@ -177,12 +177,12 @@ require (
 	k8s.io/component-base v0.0.0
 	k8s.io/cri-api v0.0.0
 	k8s.io/csi-translation-lib v0.0.0
-	k8s.io/gengo v0.0.0-20190116091435-f8a0810f38af
+	k8s.io/gengo v0.0.0-20190327210449-e17681d19d3a
```
Comment (liggitt) [inline go.mod]: also, `k8s.io/kube-openapi@v0.0.0-20190502190224-411b2483e503` requires `k8s.io/gengo@v0.0.0-20190128074634-0689ccc1d7d6`, so we'd need to bump our pinned version for that as well (hack/lint-dependencies.sh will indicate we're pinned to a backlevel, and `GO111MODULE=on go mod graph | grep gengo` shows which dependency requires the newer version)

**Pull request closed without merging**